irohad/torii/impl/command_service.cpp

```diff
@@ -155,140 +168,104 @@ namespace torii {
     return grpc::Status::OK;
   }
 
-  void CommandService::StatusStream(
-      iroha::protocol::TxStatusRequest const &request,
-      grpc::ServerWriter<iroha::protocol::ToriiResponse> &response_writer) {
-    auto resp = cache_->findItem(shared_model::crypto::Hash(request.tx_hash()));
-    if (checkCacheAndSend(resp, response_writer)) {
-      return;
-    }
-    auto finished = std::make_shared<std::atomic<bool>>(false);
-    auto subscription = rxcpp::composite_subscription();
-    auto request_hash =
-        std::make_shared<shared_model::crypto::Hash>(request.tx_hash());
-
-    /// Condition variable to ensure that current method will not return before
-    /// transaction is processed or a timeout reached. It blocks current thread
-    /// and waits for thread from subscribe() to unblock.
-    auto cv = std::make_shared<std::condition_variable>();
-
-    log_->debug("StatusStream before subscribe(), hash: {}",
-                request_hash->hex());
-
-    tx_processor_->transactionNotifier()
-        .filter([&request_hash](auto response) {
-          return response->transactionHash() == *request_hash;
-        })
-        .subscribe(
-            subscription,
-            [&, finished](
-                std::shared_ptr<shared_model::interface::TransactionResponse>
-                    iroha_response) {
-              auto proto_response = std::static_pointer_cast<
-                  shared_model::proto::TransactionResponse>(iroha_response);
-
-              log_->debug("subscribe new status: {}, hash {}",
-                          proto_response->toString(),
-                          proto_response->transactionHash().hex());
-
-              iroha::protocol::ToriiResponse resp_sub =
-                  proto_response->getTransport();
+  /**
+   * Statuses considered final for streaming. Observable stops value emission
+   * after receiving a value of one of the following types
+   * @tparam T concrete response type
+   */
+  template <typename T>
+  constexpr bool FinalStatusValue =
+      iroha::is_any<std::decay_t<T>,
+                    shared_model::interface::StatelessFailedTxResponse,
+                    shared_model::interface::StatefulFailedTxResponse,
+                    shared_model::interface::CommittedTxResponse,
+                    shared_model::interface::MstExpiredResponse>::value;
 
-              if (isFinalStatus(resp_sub.tx_status())) {
-                response_writer.WriteLast(resp_sub, grpc::WriteOptions());
-                *finished = true;
-                cv->notify_one();
-              } else {
-                response_writer.Write(resp_sub);
-              }
+  rxcpp::observable<
+      std::shared_ptr<shared_model::interface::TransactionResponse>>
```

Comment:
This type used multiple times, maybe create alias, e.g `using Response = std::shared_ptr<shared_model::interface::TransactionResponse>;`